src/bundles/arcade_2d/gameobject.ts

```diff
@@ -0,0 +1,387 @@
+import type * as types from './types';
+
+// =============================================================================
+// Classes
+// =============================================================================
+
+/**
+ * Encapsulates the basic data-representation of a GameObject.
+ */
+export abstract class GameObject implements Transformable, ReplResult {
+  private static gameObjectCount: number = 0;
+  private static gameObjectsArray: GameObject[] = []; // Stores all the created GameObjects
+  protected transformNotUpdated: boolean = true;
+  public readonly id: number;
+
+  constructor(
+    private transformProps: types.TransformProps = {
+      position: [0, 0],
+      scale: [0, 0],
+      rotation: 0,
+    },
+  ) {
+    this.id = GameObject.gameObjectCount++;
+    GameObject.gameObjectsArray.push(this);
+  }
+  setTransform(transformProps: types.TransformProps) {
+    this.transformProps = transformProps;
+    this.transformNotUpdated = true;
+  }
+  getTransform(): types.TransformProps {
+    return this.transformProps;
+  }
+  hasTransformUpdates(): boolean {
+    return this.transformNotUpdated;
+  }
+  updatedTransform() {
+    this.transformNotUpdated = false;
+  }
+
+  /**
+   * This method is called when init() Phaser Scene.
+   * @returns The GameObjects as an array.
+   */
+  public static getGameObjectsArray() {
+    return GameObject.gameObjectsArray;
+  }
+
+  public toReplString = () => '<GameObject>';
+}
+
+/**
+ * Encapsulates the basic data-representation of a RenderableGameObject.
+ */
+export abstract class RenderableGameObject extends GameObject implements Renderable {
+  protected renderNotUpdated: boolean = true;
+  private shouldBringToTop: boolean = false;
+
+  constructor(
+    transformProps: types.TransformProps,
+    private renderProps: types.RenderProps = {
+      color: {
+        red: 255,
+        blue: 255,
+        green: 255,
+        alpha: 255,
+      },
+      flip: [false, false],
+      visible: true,
+    },
+  ) {
+    super(transformProps);
+  }
+
+  setRenderState(renderProps: types.RenderProps) {
+    // if (renderProps.renderedImage === undefined || this.phaserType === undefined) {
+    //   throw new Error('GameObject\'s render image type is undefined');
+    // }
+    // if (this.phaserType in renderProps.renderedImage) {
+    //   throw new Error('Unable to update GameObject with image type that does not match');
+    // }
+    this.renderProps = renderProps;
+    this.renderNotUpdated = true;
+  }
+  getRenderState(): types.RenderProps {
+    return this.renderProps;
+  }
+  /**
+   * Sets a flag to render the GameObject infront of other GameObjects.
+   */
+  bringToTop() {
+    this.shouldBringToTop = true;
+    this.renderNotUpdated = true;
+  }
+  hasRenderUpdates(): boolean {
+    return this.renderNotUpdated;
+  }
+  updatedRender() {
+    this.renderNotUpdated = false;
+    this.shouldBringToTop = false;
+  }
+  getColor(): [number, number, number, number] {

```

Comment:
Perhaps this should be using the `ColorRGBA` type alias instead?

Reply:
Thanks for all the help!